### PR TITLE
Dump HTTP request headers

### PR DESCRIPTION
**Pull request by promag**

### CODE REVIEW

File: src/httpserver.cpp [hunk 1]
```diff
@@ -224,9 +224,13 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
     }
     std::unique_ptr<HTTPRequest> hreq(new HTTPRequest(req));
 
-    LogPrint(BCLog::HTTP, "Received a %s request for %s from %s\n",
+    LogPrint(BCLog::HTTP, "Received a %s request for %s from %s with headers:\n",
              RequestMethodString(hreq->GetRequestMethod()), hreq->GetURI(), hreq->GetPeer().ToString());
 
+    for (auto i : hreq->GetHeaders()) {
+        LogPrint(BCLog::HTTP, "  %s: %s\n", i.first, i.second);
```
Comment (promag) [inline src/httpserver.cpp]: Should redact authorization header?

**Comment by promag** [comment]
Example output:
```
2018-08-17T14:00:38Z Received a POST request for / from 127.0.0.1:61509 with headers:
2018-08-17T14:00:38Z   Authorization: Basic X19jb29raWVfXzo3OGYxMzkzMzVmNmZkNjBhZjA5ZjY3MDcyNmM0ZDE0MDRiOGI2MTlmYjYwNDc3NjExYzZmYTQ1Y2E2ZjAxODZh
2018-08-17T14:00:38Z   Connection: close
2018-08-17T14:00:38Z   Content-Length: 37
2018-08-17T14:00:38Z   Host: 127.0.0.1
```

**Comment by promag** [comment]
@practicalswift Added a commit that only dumps headers if the request is valid (authorized and method is known).

File: src/httpserver.cpp [hunk 2]
```diff
@@ -239,6 +239,12 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
         return;
     }
 
+    // Dump headers if request is valid
+    LogPrint(BCLog::HTTP, "Request headers:\n",
```
Comment (laanwj) [inline src/httpserver.cpp]: it should avoid doing this if `!LogAcceptCategory(BCLog::HTTP)`, no need for the overhead to copy the headers when the result is ignored

**Comment by laanwj** [comment]
Concept ACK, this could be useful (though is also very verbose) 

agree that string sanitizing is very important to avoid crap with escape codes etc.

**Comment by promag** [comment]
@laanwj right, I asked it above https://github.com/bitcoin/bitcoin/pull/13992#discussion_r210918337.

**Comment by laanwj** [comment]
Thinking of it, I'm not sure this is worth the added code and complexity.
It's so easy to make `tcpdump` or `wireshark` dump this information if you really need it, say for debugging your HTTP implementation.

**Comment by promag** [comment]
Updated to redact authorization header (added `EqualHeaders`).

Agree with @laanwj and I'm going to close this unless there is a strong reason to keep it.

File: src/httpserver.cpp [hunk 3]
```diff
@@ -646,6 +670,11 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod()
     }
 }
 
+bool EqualHeaders(const std::string& h1, const std::string& h2)
+{
+    return evutil_ascii_strcasecmp(h1.c_str(), h2.c_str()) == 0;
```
Comment (practicalswift) [inline src/httpserver.cpp]: Is there a reason for relying on `event2` in this function? If not I suggest doing this without relying on third-party code by introducing the required locale independent string helper functions in-tree. That would perhaps also allow us to reduce the number of know locale dependence problems listed here: https://github.com/bitcoin/bitcoin/blob/master/test/lint/lint-locale-dependence.sh#L4

**Comment by 251Labs** [comment]
Hey @promag I was looking at this PR when you closed it. 

I agree with @laanwj and you that a web debugging proxy is the right tool for inspecting HTTP requests.

If the PR was still open and you where leaning towards implementing it, I would have suggested to extract the dump headers implementation and make it a freestanding utility function.

But more importantly, IMO the changes where non-const member functions  have been changed to `const` member function are useful.

**Comment by promag** [comment]
Yeah I was going to cherry pick that.

Thanks for reviewing.

**Comment by promag** [comment]
> But more importantly, IMO the changes where non-const member functions have been changed to `const` member function are useful.

@251Labs see #14006.

**Pull request closed without merging**